### PR TITLE
Rewrite SelectEventPlugin-test to test behavior using Public API (#11299)

### PR DESCRIPTION
**Pull request by skiritsis**
I rewrote one of the two tests to use Public API, but I'm not sure about the other.

The test is supposed to verify that in case we have no event listeners attached, the events are not extracted. Isn't this decision taken internally by react though? And since they are not extracted in the first place how can you verify it without tapping into React internals? Any ideas or feedback are welcome.

### CODE REVIEW

File: packages/react-dom/src/events/__tests__/SelectEventPlugin-test.internal.js [hunk 1]
```diff
@@ -58,34 +58,36 @@ describe('SelectEventPlugin', () => {
...
+      select(event.currentTarget);
+    };
+
+    var childContainer = document.createElement('div');
+    var childNode = ReactDOM.render(
+      <input type="text" onSelect={onSelect} />,
+      childContainer,
     );
-    var node = ReactDOM.findDOMNode(rendered);
+    document.body.appendChild(childContainer);
```
Comment (gaearon) [inline packages/react-dom/src/events/__tests__/SelectEventPlugin-test.internal.js]: Do you mind doing this in `beforeEach` and then cleaning up in `afterEach`? Like [here](https://github.com/facebook/react/blob/878feebe34594fab111e47569a3a41a2f47199f1/packages/react-dom/src/events/__tests__/SyntheticClipboardEvent-test.js#L18-L30).

File: packages/react-dom/src/events/__tests__/SelectEventPlugin-test.internal.js [hunk 2]
```diff
@@ -58,34 +58,36 @@ describe('SelectEventPlugin', () => {
...
 
-    node.selectionStart = 0;
-    node.selectionEnd = 0;
+    var node = ReactDOM.findDOMNode(childNode);
```
Comment (gaearon) [inline packages/react-dom/src/events/__tests__/SelectEventPlugin-test.internal.js]: There's no need to `findDOMNode`. `render()` already gave you the right node. Old test just wasn't updated to reflect this.

File: packages/react-dom/src/events/__tests__/SelectEventPlugin-test.internal.js [hunk 3]
```diff
@@ -58,34 +68,30 @@ describe('SelectEventPlugin', () => {
...
+      container,
     );
-    var node = ReactDOM.findDOMNode(rendered);
-
-    node.selectionStart = 0;
-    node.selectionEnd = 0;
     node.focus();
+    expect(select.mock.calls.length).toBe(0);
 
-    var focus = extract(node, 'topFocus');
```
Comment (gaearon) [inline packages/react-dom/src/events/__tests__/SelectEventPlugin-test.internal.js]: This seems to be testing that a `focus` DOM event would have been ignored at this point. Shouldn't the new test also verify this?

File: packages/react-dom/src/events/__tests__/SelectEventPlugin-test.internal.js [hunk 4]
```diff
@@ -58,34 +68,30 @@ describe('SelectEventPlugin', () => {
...
-    expect(focus).toBe(null);
+    var nativeEvent = document.createEvent('Event');
```
Comment (gaearon) [inline packages/react-dom/src/events/__tests__/SelectEventPlugin-test.internal.js]: Let's use the newer API for this? e.g. `new MouseEvent('mousedown', {bubbles: true, cancelable: true})`

**Comment by gaearon** [comment]
Is something blocking you from changing the second test?

**Comment by skiritsis** [comment]
@gaearon Thank you for taking the time to review the changes. Like I mentioned on the description, the other test that has not been changed is supposed to "skip extraction if no listeners are present". This means that in order to test it we should not add any React event listeners on the element and make sure that the events were not emitted. Since React though internally does not emit the event if the listener is not present how can we verify the behavior without accessing its internals?

**Comment by gaearon** [comment]
The test was originally written in 47b147f392450985bc7bd38d98a4c9e1e7188150 (found it via blame). The commit description describes the issue that it was trying to solve:

>For instance, if `topMouseDown` has been registered but not `topMouseUp`, event extraction will set the `mouseDown` flag to true but never unset it. This leads to bugs when `onSelect` is registered and should be firing during normal key behavior. Since no `topMouseUp` has yet occurred to unset the flag, `onSelect` fails to fire.

Perhaps you can manage to reproduce this exact problem in a new test? You can then "undo" some parts of the fix in 47b147f392450985bc7bd38d98a4c9e1e7188150 to verify your code still tests it.

**Comment by skiritsis** [comment]
Thank you again. I will try to reproduce it with a different test and let you know.

**Comment by gaearon** [comment]
Please let me know if you get stuck!

**Comment by skiritsis** [comment]
Sorry for the late response, I rewrote the other test as well. It may look weird and I could be way wrong here, but l'm looking at the code path and it seems to me that is exactly the same.

**Comment by gaearon** [comment]
If I comment out [this line](https://github.com/facebook/react/blob/c3f1b6cd91ffa3e05465c3ed2f48225876c1bd3b/packages/react-dom/src/events/SelectEventPlugin.js#L152) on master, the test fails. But it doesn't fail in this branch. Doesn't this mean we're no longer verifying that behavior?

**Comment by gaearon** [comment]
I think what https://github.com/facebook/react/pull/3639 says is not that `mouseup` didn't fire, but that it fired but there was no *React listener* for it (and thus SelectEventPlugin didn't "see" this even). React attaches top-level listeners lazily.

File: packages/react-dom/src/events/__tests__/SelectEventPlugin-test.internal.js [hunk 5]
```diff
@@ -11,81 +11,81 @@
...
+    });
+    node.dispatchEvent(nativeEvent);
+
+    nativeEvent = new MouseEvent('mousedown', {
+      bubbles: true,
+      cancelable: true,
+    });
+    node.dispatchEvent(nativeEvent);
+
+    nativeEvent = new MouseEvent('keyup', {bubbles: true, cancelable: true});
```
Comment (gaearon) [inline packages/react-dom/src/events/__tests__/SelectEventPlugin-test.internal.js]: Note: keyup is not a mouse event.

**Comment by gaearon** [comment]
I think the correct test would be:

```js
  // See https://github.com/facebook/react/pull/3639 for details.
  it('does not get confused when dependent events are registered independently', () => {
    var select = jest.fn();
    var onSelect = event => select(event.currentTarget);

    // Pass `onMouseDown` so React registers a top-level listener.
    var node = ReactDOM.render(
      <input type="text" onMouseDown={function() {}} />,
      container,
    );
    node.focus();

    // Trigger `mousedown` and `mouseup`. Note that
    // React is not currently listening to `mouseup`.
    node.dispatchEvent(
      new MouseEvent('mousedown', {
        bubbles: true,
        cancelable: true,
      }),
    );
    node.dispatchEvent(
      new MouseEvent('mouseup', {
        bubbles: true,
        cancelable: true,
      }),
    );

    // Now subscribe to `onSelect`.
    ReactDOM.render(<input type="text" onSelect={select} />, container);
    node.focus();

    // This triggers a `select` event in our polyfill.
    node.dispatchEvent(
      new KeyboardEvent('keydown', {bubbles: true, cancelable: true}),
    );
    node.dispatchEvent(
      new KeyboardEvent('keyup', {bubbles: true, cancelable: true}),
    );

    // Verify that it doesn't get "stuck" waiting for
    // a `mouseup` event that it has "missed" because
    // a top-level listener didn't exist yet.
    expect(select.mock.calls.length).toBe(1);
  });
```

**Comment by skiritsis** [comment]
Thanks for editing. I believe this
 

> but that it fired but there was no React listener for it (and thus SelectEventPlugin didn't "see" this even). React attaches top-level listeners lazily.

made it a lot more clear to me.

**Comment by gaearon** [comment]
Yea, took me a bit of time to figure out myself. Reading what `SelectEventPlugin` actually does helped.

**Pull request merged**